### ISSUE TITLE
UX improvements discussion

### ISSUE THREAD

**Issue by majkinetor** [VShawn/PRemoteM#34]
1. Allow multiselect of boxes and bulk options:
    - delete
    - move to group (or drag and drop to tab)
    - export config
    - config edit (all options except `Name` & `Address`).
1. List view as current box only view isn't practical. Launcher mitigates that but here you would want to organize stuff.
2. ESC to cancel any option UI
3. Typing in main UI should go to search
4. Current hovering animation is annoying IMO (maybe better to glow or something less intrusive)
1. There is no need to save Logo separately from entire config. Its not intuitive and I missed the one place or another several times already.
1. Purpose of color ? I don't see any difference. Color is otherwise awesome to mark dev/stage/production and several 
    products do that so (unless you are full screen) you always know how critical environment is by the border and background  colors on various places (this could affect main view, launcher view, tab color etc.) <details><summary>see [dbeaver ](https://github.com/dbeaver/dbeaver)for inspiration (red prod, yellow stage, white devel)</summary> 
    ![image](https://user-images.githubusercontent.com/85767/101668967-8d297b00-3a51-11eb-9d64-934d98f88ed1.png)
    </details>
1. Remember window maximization state
2. Allow middle click to close tab (standard) or add x like Chrome.

**Comment by VShawn** [comment]
I had a plan for these, the operation logic just like what you list out. AMOF I just stuck at interact  design like UI and animation since I know few about design things.

**Comment by VShawn** [comment]
>> Current hovering animation is annoying

sorry, i not sure **hovering animation** means which animation?
Logo zooming? or box border change?

**Comment by majkinetor** [comment]
Sorry didnt even notice the border change :) Logo zooming. Not to mention that logo zoomed looks pixelated.

**Comment by Shooshka** [comment]
UX related I think... tranparent isn't good for all pictures ;)

![2020-12-09_18-53-01](https://user-images.githubusercontent.com/1889052/101652964-ea243180-3a4f-11eb-85d2-166e97919a3d.png)

**Comment by majkinetor** [comment]
Added 2 more items (logo save and color)

**Comment by VShawn** [comment]
> Purpose of color 

In the very begining, there was a color border around session work area, use to identify sessions.

but now color only shown on tab since I found it ugly to have a color border around.


![image](https://user-images.githubusercontent.com/10143738/101731087-0e3a4e00-3af6-11eb-9506-b765df8a8b68.png)

**Comment by VShawn** [comment]
<details><summary> I made a demo page for management, just for feasibility analysis</summary> 

![image](https://user-images.githubusercontent.com/10143738/101917125-d8cf5680-3c02-11eb-9639-7c7bfadef944.png)

</details>

**Comment by majkinetor** [comment]
Looks great 🥇

**Comment by profbaco** [comment]
This idea of seeing items in list form is interesting! I was going to open an issue for this myself.
========================
Esta ideia de ver os itens em forma de lista é em interessante! Eu mesmo ia abrir uma issue para isto.

**Comment by majkinetor** [comment]
Updated issue with 8.

1. Option to remove titlebar and/or tabs to achive bigger vertical space <details><summary>details</summary> 
![image](https://user-images.githubusercontent.com/85767/102208945-60051e80-3ed0-11eb-8eed-43219a088a65.png)</details>

**Comment by VShawn** [comment]
> Updated issue with 8.
 
have fun with #28  🌝

**Comment by majkinetor** [comment]
OK removed that one, added new one:

1. Remeber window state (maximized, window size)


The point here is this - app starts remote connection very fast, so I don't keep them around, i just close them and relaunch them when I need it again - this is what I prefer since while you are on the remote, you cant easily alt tab with your host and when alt tab enters the remote, you cant easily go out to host particularly on full screen.

For that rason I keep remotes windowed or maximized which resets always when I close the main app window. Since resizing takes time (at least 2-3s) it would be way better to remember.

**Comment by VShawn** [comment]
## size remember

<details><summary>It did remember lasest tab window size</summary> 

![image](https://user-images.githubusercontent.com/10143738/102232892-9d969580-3f2a-11eb-8741-1d54a2a70ae1.png)

</details>

But the point is: in tab mode window, new remote session will be  attached into the on tab.

So if we open a 200 X 100 remote session A first, then try to open a 300 X 100 remote B, this B session will not open with size 300 X 100, but just join the Tab of session A and become a 200 X 100 size session. I dont know how to handle these above, so  I just leave it alone when i build tab module.

## maximized state are not remember
I will mark this and add maximized state remember next turn.

**Comment by majkinetor** [comment]
> So if we open a 200 X 100 remote session A first, then try to open a 300 X 100 remote B, this B session will not open with size 300 X 100, but just join the Tab of session A and become a 200 X 100 size session. I dont know how to handle these above, so I just leave it alone when i build tab module.

Yeah, that only makes a problem with resize mode fixed. And to jump sizes when you over tabs, thats retarded and won't work well anyway. So perhaps this setting should be removed ? Or to document that it only works if there are no other active connections.

**Comment by majkinetor** [comment]
Added 

1. Allow middle click to close tab (standard) or add x like Chrome.

**Comment by VShawn** [comment]
> UX related I think... tranparent isn't good for all pictures ;)
> 
> ![2020-12-09_18-53-01](https://user-images.githubusercontent.com/1889052/101652964-ea243180-3a4f-11eb-85d2-166e97919a3d.png)

fix by make icon small

![image](https://user-images.githubusercontent.com/10143738/102754255-a31b3200-43a7-11eb-9814-95b610e024fb.png)

**Comment by VShawn** [comment]
> Added
> 
> 1. Allow middle click to close tab (standard) or add x like Chrome.

![image](https://user-images.githubusercontent.com/10143738/102787489-1dfe4000-43dc-11eb-9c62-9342ce22e4eb.png)

**Comment by majkinetor** [comment]
Few observations about latest release 0.5.7 which 

### Edit selection

Why didn't u reuse existing UI to create connection. You could just gray fields that are not changable in group such as Name (its strange you left it). <details><summary>details</summary> 

Fields that are crossed should be disabled in group edit:

![image](https://user-images.githubusercontent.com/85767/103097038-3a7dc080-4606-11eb-8b4d-0904384c8ce5.png)
</details> 

### List  

- Is it possible to resize font ? Maybe without an option, using standard `ctrl +|-` or mouse ?

### Icons 

- You should probably add more general icons (or better, support something like  [font awesome](https://fontawesome.com/) then we have hundreedss). Current ones are too much vendor specific.

**Comment by majkinetor** [comment]
> You could just gray fields that are not changable in group

Or if you go with your current approach, you could totally ignore this and make ALL fields settable in group, if that is what user wants for some reason. 

Its enough that name and address are similar for this feature to be usable, for example have the same prefix which you could then set to entire group and edit each item to add suffix.

**Comment by VShawn** [comment]
@majkinetor 

## Edit selection
no no, current approach not  what i wanted,  it just easy to implement for now software architecture.

and... Visual Studio  approach  is what I wanted, in case below crossed fields would show as **'\<different options\>'**, but this needs re-design our code.

![image](https://user-images.githubusercontent.com/10143738/103112292-62fdcd00-468f-11eb-9582-bc3daa357255.png)

## List
actually I don‘t know how to control global font size dynamically : ( , do you think font size is too small not?

## server Icons/logo
I am going to add more icons, but I am not sure about the commercial authorization of the icons and I lost my source server logo design files for some reason. In addition, I feel that the style of *font awesome* doesn't match the current one.

**Comment by VShawn** [comment]
we add tons of icons 😸 


![image](https://user-images.githubusercontent.com/10143738/103768347-4bdeb800-505d-11eb-9114-7fce6bbcd765.png)

**Comment by VShawn** [comment]
#69  Bug report form UX needs improvement

**Comment by NAV-Management** [comment]
HI @VShawn,

thank you very much for the RDP tool. For me I already feel that it is a good and useful RDP tool. I have tried several RDP tools in the past, but none of the tools I have used as long as the outdated RDP Connection Manger, because it is no longer available. and the other tools I have quickly removed again because I always had something wrong with the tools.

I've been testing your tool for 2 days now and I'm very very happy about the dynamic customization of RDP resolution, that's a feature that no free tool I've had my hands on so far has had.

I would love to bring some suggestions:

- When an RDP connection is active, there should be an indicator that shows in the main menu that the connection is currently active. 
Possibilities of an indicator: Either the text is changed in color, for example to green, or a check mark is added to the logo. 
Or there is another column that points to it, but I personally find unnecessary with another column, because this can be built in by other ways.
If this would be built in color wise, you could integrate this into the themes section right away :).
- This can be further spun at the point that the text becomes red, for example, if the connection was terminated for some reason or was not established. Or that a red cross is displayed on the logo.
- Add another button in the context menu of the server management to open all marked connections.
- In the RDP bar above you can see the icon on the left side of the active RDP connections, please make it bigger, because I can't see any icon there, except a weird pixel blob.
As a Windows window icon, however, it works wonderfully, and here again thank you that there is such a feature with logo at all.
- It would still be desirable that in the Quick-Connect the possibility is built in to also search for the groups, so that the group titles can also be searched.


Regarding the performance, I noticed two things:

- Searching for connections in the main menu is smooth, but if you search the RDP list in Quickconnect, you notice that it hangs significantly. 
We have almost 60 RDP connections in the list to see that in relation.

- If an RDP connection is disconnected, the session remains in RAM. In my way of working I often connect a lot to RDP servers and also disconnect / log off. Most of the time I "only" have 3 to 4 RDP connections active, but due to the closed RDP connections the RAM consumption grows to 2.5GB so far. If necessary, one can still optimize here that the RAM is released further. For me, the RDP software remains active most of the time, I try to avoid having to restart it. 
Since the RDP manager of CINSPIRATION was 3x worse. There it went up to 5GB RAM.

I know it was already mentioned in other tickets, but in the workflow it helps a lot to have a tree structure in the RDP tool here, instead of a tab structure. Since IMHO more data can be displayed in the tree than a side tab structure.

Thanks for reading and for providing a new / fresh RDP tool :).

**Comment by majkinetor** [comment]
> Searching for connections in the main menu is smooth, but if you search the RDP list in Quickconnect, you notice that it hangs significantly.

I can't confirm this. I have around 50 connections and quick launcher works instantly.

> If an RDP connection is disconnected, the session remains in RAM

This was reported previously (#37) and after it was fixed I didn't have any problems - as soon as RDP connection is closed RAM is released.

**Comment by NAV-Management** [comment]
About QuickConnect:
> I can't confirm this. I have around 50 connections and quick launcher works instantly.

In other words, search works fine, but when you delete text in the search bar, the delay before the letter is deleted is noticeable.

About RAM:
> This was reported previously (#37) and after it was fixed I didn't have any problems - as soon as RDP connection is closed RAM is released.
So I have completely closed the PRemoteM and reopened it. Then I started 5 RDP sessions and ended them after a short time. In conclusion, 1.8GB RAM is still needed although ~ 70MB RAM is devoured after starting the tool.

> To workaround this, I name all my connections with group name as prefix which is simple and easy thing to do. I think you didn't express it right tho - I guess you want to , by typing the group name, PRM to show all connections in that group even if they don't contain that phrase in the name. It looks like you want to show the group itself, and what will it mean when selected ? show submenu ? Not very quicky and defeats the spirit of quick launcher. But first mentioned case makes sense (as an option ofcourse)

Yes, of course I could rename all my connections, and put the group name as a prefix. But then the names will be much longer than they already are and otherwise the question is, why there are groups at all, if you can't use that as a filter criterion. 

In my case I have almost 20 groups, where a maximum of 5 servers are stored per group and then it makes IMHO sense to search the group names in the Quickconnect, and quickly find the appropriate server and open. Now and then I would also open all found servers via the way, because I have to work in parallel on all servers of the group.

Another point I found, from time to time I use a secondary account to connect to the server. So far I have opened a classic RDP connection to the server as a workaround. As an alternative, there would be the possibility to create the server twice and store the other credentials there, but makes the whole thing not really comfortable. 

This point I found the mRemoteNG completely idiotic solved, or not at all. Since at the point not even the inheritance of the account data worked properly. Since I am then very quickly changed.


Although it looks like I'm writing badly about the tool. No, I think it is really well done and does a lot different and better than other RDP tools. Only these are points that I miss and I wish that they are integrated and therefore I am also looking forward to the future of the tool.

I can program C# myself, unfortunately only with WinForms, otherwise if I had time, I would like to support if this is desired :).

**Comment by VShawn** [comment]
Hi @NAV-Management thank your for your valuable advice, sorry for relpy late, I just so busy these days 😵 

> indicator 

I will think about how to add indicator 😉 

> In the RDP bar above you can see the icon on the left side of the active RDP connections, please make it bigger

errrrr which icon do you mean? 
this one?
![image](https://user-images.githubusercontent.com/10143738/108186075-ed225900-7147-11eb-85bc-07870156b61d.png)

give a screen shoot would be good.


> Option to automatically close the RDP tab after logging out from the server side.

no sure if we can get the log out even from server side, i need to check the document.

> Yes, of course I could rename all my connections, and put the group name as a prefix. But then the names will be much longer than they already are and otherwise the question is, why there are groups at all, if you can't use that as a filter criterion.

You are right, group name should take place in the launcher, the reason why it not in launcher is that in the very original design(v0.0.1), there was no launcher for PRM. I add the launcher in v 0.1.0 : )

> Searching for connections in the main menu is smooth, but if you search the RDP list in Quickconnect, you notice that it hangs significantly.

The point is I designed Quickconnect for myself, about 20 connections and it works very well, 60 connections was never been tested 😓.  IMO hangs in the launcher is probably because:
- the performance of WPF UI, every time you changed keyword will refresh the listview, this may hangs the UI. I'm not good at UI so we need some time to find out how to improve it.
- the performance of search algorithm, I try to fix this by #74 


> RAM

For me, there is no idea to solve this problem temporarily. On my machine, the memory will be released within a period of time (a few secondes, 10s 20s) after closing the session, but it cannot be released completely


|  | after start | after rdp session open|after session colose|
| :-----| :---- | :---- | :---- |
| RAM | 25MB | 300+MB |70MB |


> I can program C# myself, unfortunately only with WinForms, otherwise if I had time, I would like to support if this is desired :).

support are always welcome, WPF is easy to learn(IMO), I program C++ on work, C#/WPF is my emmm.... spare time entertainment, I am not very good at it.


thanks again for your valuable advices

**Comment by NAV-Management** [comment]
I found another idea for an indicator:
Add a server type option field that could include fields like "Productive", "Testing", "Developing". 
And in the theme settings, add a color range for the server types. 

So in the example, if you set green as the color for Develop, blue for Test, and red for Productive. When you connect to a server, the color of the tab changes according to the server type it is. This way you can see at a glance where you are.

Regarding the logo topic, I mean that a self-made logo is hardly displayed in the tab: 
![image](https://user-images.githubusercontent.com/24463981/108195448-0f03e780-7118-11eb-9d57-97b56f48a9ed.png)
Can you tell me what size the logo must be?

Regarding automatic closing the Tab, their should be documentation about it, because most rdp-tools can close it:
In Example RDP-Manager from Cinspiration:
Connect: 
![image](https://user-images.githubusercontent.com/24463981/108195922-c698f980-7118-11eb-9730-f1ba4149bea2.png)
Successfully logged in:
![image](https://user-images.githubusercontent.com/24463981/108196031-e7614f00-7118-11eb-8f4a-798f9d9b07f6.png)
Manually Logged off from Server:
![image](https://user-images.githubusercontent.com/24463981/108196147-0a8bfe80-7119-11eb-9bc3-0f0e11db7087.png)

Regarding RAM: yes most of the tools have the problem. But the RDP manager consumes a lot of RAM after some working time.
Regarding WPF programming: The GUI is very unique in this area, and I find it quite good

**Comment by NAV-Management** [comment]
Sorry if I disturb you, but i found another bug:
When you log out of a Win 2008 R2 server, you will be logged back in automatically if you are not fast enough to close the tab.

**Comment by majkinetor** [comment]
>  but i found another bug:

@NAV-Management please create a new ticket for it.

>  Now and then I would also open all found servers via the way, because I have to work in parallel on all servers of the group.

Ah, I guess nobody envisioned that. And it could really be a nice feature but it needs more serious design as some users may not want that. I personally think the idea has merit, pattern you described is not that uncommon and it is similar to a way some shell tools allow predefinition of apps that should be run together (tmux comes to mind, but also ConEmu on Windows).  Gimme some time to think about it and I will create a ticket with detailed description. This will probably end up in list of ideas (Limbo milestone here). Those ideas can then be voted, analyzed more or even PRed by individuals that do have extra time.

> Another point I found, from time to time I use a secondary account to connect to the server. So far I have opened a classic RDP connection to the server as a workaround. As an alternative, there would be the possibility to create the server twice and store the other credentials there, but makes the whole thing not really comfortable.

Not sure what is your proposition here. When I have this problem I create copy of existing one and change the user. Do you mean that you would like to have some kind of extra menu that shows the list of users. If so, it could be abstracted to any provider, not just RDP. The benefit is probably miniscule anyway, but implementation too, and anything leading to more maintainable list of connections is probably good to have given that it is the main purpose of the tool.

> Since at the point not even the inheritance of the account data worked properly.

Inheritance is buggy/doesn't-work in MRNG and its very needed indeed. I mentioned group defaults at #73 which is the same as inheritance.

> I can program C# myself, unfortunately only with WinForms, otherwise if I had time, I would like to support if this is desired :).

We all have our RL duties, I would also personally really like to get a stab at most of the ideas proposed I wouldn't want however to see PRM going into ad hoc direction especially given the limited dev man power of N=1 ATM its best not to overload it with demands, but nothing is wrong with hoarding them as situation might change any day.

**Comment by majkinetor** [comment]
> The point is I designed Quickconnect for myself, about 20 connections and it works very well, 60 connections was never been tested

Don't be ridiculous @VShawn :) We are not talking about 50K connections, 20, 60 or 500 should be all the same really.

**Comment by majkinetor** [comment]
> Add a server type option field that could include fields like "Productive", "Testing", "Developing".

See point 7 about dbeaver above.

**Comment by NAV-Management** [comment]
@majkinetor,

thanks for the detailed reply :).

> @NAV-Management please create a new ticket for it.

I created the ticket #82.

> Ah, I guess nobody envisioned that. And it could really be a nice feature but it needs more serious design as some users may not want that. I personally think the idea has merit, pattern you described is not that uncommon and it is similar to a way some shell tools allow predefinition of apps that should be run together (tmux comes to mind, but also ConEmu on Windows). Gimme some time to think about it and I will create a ticket with detailed description. This will probably end up in list of ideas (Limbo milestone here). Those ideas can then be voted, analyzed more or even PRed by individuals that do have extra time.

Will be interesting how it changes the APP :).

> Not sure what is your proposition here. When I have this problem I create copy of existing one and change the user. Do you mean that you would like to have some kind of extra menu that shows the list of users. ....

Yeah I mean something like an extra menu or an entry in the conext menu.

> We all have our RL duties, I would also personally really like to get a stab at most of the ideas proposed I wouldn't want however to see PRM going into ad hoc direction especially given the limited dev man power of N=1 ATM its best not to overload it with demands, but nothing is wrong with hoarding them as situation might change any day.

I know it and didn't want to change to much on it, just things I found out :)
On my side, I forked it and added german language in it. But german is on some words very long, so the text get cut off and I don't know how I could short it much more that it fits.
Yellow marked text, get cut off:
![image](https://user-images.githubusercontent.com/24463981/108289838-74df8600-718f-11eb-94e7-d9d722e11d50.png)
![image](https://user-images.githubusercontent.com/24463981/108289856-7dd05780-718f-11eb-9d66-8cf8067e8445.png)
![image](https://user-images.githubusercontent.com/24463981/108289888-89bc1980-718f-11eb-8f50-a7e7668868d2.png)


> See point 7 about dbeaver above.
Oops I had overlooked that then

**Comment by majkinetor** [comment]
Thanks @NAV-Management, I created #83 to discuss multi conn. options.

**Comment by VShawn** [comment]
@NAV-Management 

> Can you tell me what size the logo must be? 

it is better using 100 × 100 pix image.

> But german is on some words very long, so the text get cut off and I don't know how I could short it much more that it fits.

Same problem when I try translate it into Japanese. So I wanna re-build UI with Net5 + https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit

the field may change like blow so as to getting enough space for description.

![image](https://user-images.githubusercontent.com/10143738/108336206-bfecae00-720e-11eb-8ba2-0058db0f5864.png)

![image](https://user-images.githubusercontent.com/10143738/108336332-e1e63080-720e-11eb-992f-207f33b77ae4.png)

**Comment by majkinetor** [comment]
Mot of the items here are either already implemented or have their own ticket so I will close this.
